### PR TITLE
Added LIBC_LOCALTIME support for date command

### PR DESCRIPTION
**Pull request by thapav**
Added Localtime support:
Date command works fine with CONFIG_LIBC_LOCALTIME enabled.

Signed-off-by: Vidisha Thapa <thapa.v@samsung.com>

### CODE REVIEW

File: apps/system/utils/kdbg_date.c [hunk 1]
```diff
@@ -127,6 +127,9 @@ static inline int date_settime(int argc, char **args)
 	FAR char *saveptr;
 	long result;
 	int ret = OK;
+#ifdef CONFIG_LIBC_LOCALTIME
+	tm.tm_isdst = 0;
```
Comment (manoharahk) [inline apps/system/utils/kdbg_date.c]: Is it not required to parse the parameters of date_settime and set all the members associated with CONFIG_LIBC_LOCALTIME) || CONFIG_TIME_EXTENDED accordingly?

-----
struct tm {                                                                      
        int tm_sec;                                     /* second (0-61, allows for leap seconds) */
        int tm_min;                                     /* minute (0-59) */         
        int tm_hour;                            /* hour (0-23) */                
        int tm_mday;                            /* day of the month (1-31) */       
        int tm_mon;                                     /* month (0-11) */          
        int tm_year;                            /* years since 1900 */           
#if defined(CONFIG_LIBC_LOCALTIME) || defined(CONFIG_TIME_EXTENDED)              
        int tm_wday;                            /* day of the week (0-6) */         
        int tm_yday;                            /* day of the year (0-365) */       
        int tm_isdst;                           /* non-0 if daylight savings time is in effect */
#endif                                                                           
};

**Comment by sunghan-chang** [comment]
@thapav Your commit title is what problem is. Let's change it to what you do.

File: apps/system/utils/kdbg_date.c (hunk 1 above)
Comment (thapav) [inline apps/system/utils/kdbg_date.c]: initialized the local copy all the members associated with CONFIG_LIBC_LOCALTIME) || CONFIG_TIME_EXTENDED and then set coresponding values

**Comment by thapav** [comment]
Please follow latest date command format in case of CONFIG_LIBC_LOCALTIME enabled.
Follow comments for the same.

**Comment by thapav** [comment]
Please review changes

File: apps/system/utils/kdbg_date.c [hunk 2]
```diff
@@ -247,6 +302,21 @@ int kdbg_date(int argc, char **args)
...
 		ret = ERROR;
+#else
+		printf("\nUsage: date\n");
+		printf("   or: date [-s FORMAT]\n");
+		printf("Display, or Set system time and date information\n");
+		printf("\nOptions:\n");
+		printf(" -s FORMAT     Set system time in the given FORMAT\n");
+		printf("               FORMAT: MMM DD HH:MM:SS YYYY DWK DYR D\n");
+		printf("                day of week range = [0,6] , (Sunday = 0)\n");
+		printf("                day of week range = [0,365], \n");
```
Comment (sunghan-chang) [inline apps/system/utils/kdbg_date.c]: 1. day of year range?
2. [0, 365] means from 0 to 365. Isn't it from 0 to 364?

File: apps/system/utils/kdbg_date.c [hunk 3]
```diff
@@ -247,6 +302,21 @@ int kdbg_date(int argc, char **args)
...
+		printf("                daylight_savings_flag: 1(high), 0(low) \n");
+		printf("                'month', 'day', 'hour':'minute':'second' 'year' 'day_of_week' 'day_of_year' 'daylight_savings_flag'\n");
+		printf("                Example: Apr 21 10:35:22 1991 Sun 175 1\n");
```
Comment (sunghan-chang) [inline apps/system/utils/kdbg_date.c]: Why do we need to put "day of year" even though MMM DD is there?

File: lib/libc/time/lib_strftime.c [hunk 4]
```diff
@@ -191,15 +191,13 @@ size_t strftime(FAR char *s, size_t max, FAR const char *format, FAR const struc
 		switch (*format++) {
 #if defined(CONFIG_LIBC_LOCALTIME) || defined(CONFIG_TIME_EXTENDED)
 		/* %a: A three-letter abbreviation for the day of the week. */
-
```
Comment (sunghan-chang) [inline lib/libc/time/lib_strftime.c]: Why do you remove these two lines?

**Comment by thapav** [comment]
@sunghan-chang 
1. day of year range includes leap year days as well, hence 0-365
2. Day of the year is a localtime variable which we need to provide to tm struct if LIBC_LOCALTIME  is enabled

File: apps/system/utils/kdbg_date.c [hunk 5]
```diff
@@ -235,9 +289,10 @@ int kdbg_date(int argc, char **args)
 
 	if (argc == 1) {
 		ret = date_showtime();
-	} else if (!strcmp(args[1], "-s") && argc == 6) {
+	} else if (!strcmp(args[1], "-s") && argc == 9) {
```
Comment (sunghan-chang) [inline apps/system/utils/kdbg_date.c]: In this case, #if !(defined(CONFIG_LIBC_LOCALTIME) || defined(CONFIG_TIME_EXTENDED)), argc is 6. But you modify this without any conditional.

File: apps/system/utils/kdbg_date.c [hunk 6]
```diff
@@ -90,9 +94,24 @@ static inline int date_month(FAR const char *abbrev)
...
+	}
+
+	return ERROR;
+}
+#endif
+
 static inline int date_showtime(void)
 {
-	static const char format[] = "%b %d %H:%M:%S %Y";
+	static const char format[] = "%b %d %H:%M:%S %Y %a";
```
Comment (sunghan-chang) [inline apps/system/utils/kdbg_date.c]: It looks not related with commit message.

File: apps/system/utils/kdbg_date.c [hunk 7]
```diff
@@ -247,6 +302,21 @@ int kdbg_date(int argc, char **args)
 		printf("                'month', 'day', 'hour':'minute':'second' 'year'\n");
 		printf("                Example: Apr 21 10:35:22 1991\n");
 		ret = ERROR;
+#else
+		printf("\nUsage: date\n");
+		printf("   or: date [-s FORMAT]\n");
+		printf("Display, or Set system time and date information\n");
+		printf("\nOptions:\n");
+		printf(" -s FORMAT     Set system time in the given FORMAT\n");
+		printf("               FORMAT: MMM DD HH:MM:SS YYYY DWK DYR D\n");
```
Comment (sunghan-chang) [inline apps/system/utils/kdbg_date.c]: 1. I guess DYR is "Day of Year Range". But why "Day of Week Range" is DWK?
2. In date_settime, you compare string. But here it looks we can use a number, not string.

File: apps/system/utils/kdbg_date.c [hunk 8]
```diff
@@ -90,9 +94,24 @@ static inline int date_month(FAR const char *abbrev)
 	return ERROR;
 }
 
+#if defined(CONFIG_LIBC_LOCALTIME) || defined(CONFIG_TIME_EXTENDED)
+static inline int day_of_week(FAR const char *abbrev)
+{
+	int wday;
+
+	for (wday = 0; wday < 7; wday++) {
+		if (strncasecmp(g_dayofweek[wday], abbrev, 3) == 0) {
```
Comment (sunghan-chang) [inline apps/system/utils/kdbg_date.c]: Will you use string? or number? If you will use string, why do you write like below?
```
printf("                day of week range = [0,6] , (Sunday = 0)\n");
```

File: apps/system/utils/kdbg_date.c [hunk 9]
```diff
@@ -205,6 +224,41 @@ static inline int date_settime(int argc, char **args)
...
+	}
+
+	tm.tm_wday = day_of_week(token);
+        if (tm.tm_wday < 0) {
+		goto errout_bad_parm;
+	}
+
+	/* Get the day of the year.  NOTE: Accepts day-of-year from 0 to 365 */
+
+	token = args[6];
```
Comment (sunghan-chang) [inline apps/system/utils/kdbg_date.c]: 7, not 6.

File: apps/system/utils/kdbg_date.c [hunk 10]
```diff
@@ -205,6 +224,41 @@ static inline int date_settime(int argc, char **args)
...
+	if (token == NULL) {
+		tm.tm_yday = 0;
+	}
+	else {
+		tm.tm_yday = (int)token;
+	}
+
+	/* Set the daylight savings flag value */
+
+	token = args[6];
```
Comment (sunghan-chang) [inline apps/system/utils/kdbg_date.c]: 8, not 6.

**Comment by thapav** [comment]
@sunghan-chang 
modified as per review comments

File: apps/system/utils/kdbg_date.c [hunk 11]
```diff
@@ -235,9 +299,10 @@ int kdbg_date(int argc, char **args)
 
 	if (argc == 1) {
 		ret = date_showtime();
-	} else if (!strcmp(args[1], "-s") && argc == 6) {
+	} else if (!strcmp(args[1], "-s") && (argc >= 6 || argc <= 9)) {
```
Comment (sunghan-chang) [inline apps/system/utils/kdbg_date.c]: When we give 7 arguments, it will call "date_settime". Is it working?

File: apps/system/utils/kdbg_date.c [hunk 12]
```diff
@@ -247,6 +312,20 @@ int kdbg_date(int argc, char **args)
...
 		printf("                Example: Apr 21 10:35:22 1991\n");
 		ret = ERROR;
+#else
+		printf("\nUsage: date\n");
+		printf("   or: date [-s FORMAT]\n");
+		printf("Display, or Set system time and date information\n");
+		printf("\nOptions:\n");
+		printf(" -s FORMAT     Set system time in the given FORMAT\n");
+		printf("               FORMAT: MMM DD HH:MM:SS YYYY DWR DYR D\n");
+		printf("                day of week range = [Sunday = Sun, Monday = Mon, and so on...]\n");
```
Comment (sunghan-chang) [inline apps/system/utils/kdbg_date.c]: How about change first character on each word to match easily with above FORMAT?
```
Day of Week Rage (DWR) = xxxxx
```

File: apps/system/utils/kdbg_date.c (hunk 11 above)
Comment (thapav) [inline apps/system/utils/kdbg_date.c]: yes

File: apps/system/utils/kdbg_date.c (hunk 12 above)
Comment (thapav) [inline apps/system/utils/kdbg_date.c]: sounds good
modified it

File: apps/system/utils/kdbg_date.c [hunk 13]
```diff
@@ -205,6 +228,47 @@ static inline int date_settime(int argc, char **args)
...
+	}
+
+	/* Get the day of the year.  NOTE: Accepts day-of-year from 0 to 365 */
+
+	token = args[7];
+	if (token == NULL) {
+		tm.tm_yday = 0;
+	}
+
+	result = strtol(token, NULL, 10);
```
Comment (sunghan-chang) [inline apps/system/utils/kdbg_date.c]: If token is null, you gave zero for tm_yday at line 248. But due to line, that is ignored. Let's remove line 248 or add "else".

File: apps/system/utils/kdbg_date.c [hunk 14]
```diff
@@ -205,6 +228,47 @@ static inline int date_settime(int argc, char **args)
...
+	tm.tm_yday = (int)result;
+
+	/* Set the daylight savings flag value */
+
+	token = args[8];
+	if (token == NULL) {
+		tm.tm_isdst = 0;
+	}
+
+	result = strtol(token, NULL, 10);
```
Comment (sunghan-chang) [inline apps/system/utils/kdbg_date.c]: same as previous

File: apps/system/utils/kdbg_date.c [hunk 15]
```diff
@@ -235,9 +299,10 @@ int kdbg_date(int argc, char **args)
 
 	if (argc == 1) {
 		ret = date_showtime();
-	} else if (!strcmp(args[1], "-s") && argc == 6) {
+	} else if (!strcmp(args[1], "-s") && (argc >= 6 && argc <= 9)) {
```
Comment (sunghan-chang) [inline apps/system/utils/kdbg_date.c]: When CONFIG_LIBC_LOCALTIME is enabled and argc is given 6, date_settime is called. But becuase of line 236, it goes failure always.

**Comment by sunghan-chang** [comment]
And your commit title is not enough to know what commit is so that I should watch the changes to know what it is. Could you modify titles?

File: apps/system/utils/kdbg_date.c (hunk 15 above)
Comment (thapav) [inline apps/system/utils/kdbg_date.c]: oh my mistake, it should be 7 to 9

File: apps/system/utils/kdbg_date.c (hunk 13 above)
Comment (thapav) [inline apps/system/utils/kdbg_date.c]: done

File: apps/system/utils/kdbg_date.c (hunk 14 above)
Comment (thapav) [inline apps/system/utils/kdbg_date.c]: done

File: apps/system/utils/kdbg_date.c (hunk 15 above)
Comment (sunghan-chang) [inline apps/system/utils/kdbg_date.c]: You should concern conditionals without disabled LOCALTIME also.

File: apps/system/utils/kdbg_date.c [hunk 16]
```diff
@@ -205,6 +228,51 @@ static inline int date_settime(int argc, char **args)
...
+		goto errout_bad_parm;
+	}
+
+	/* Get the day of the year.  NOTE: Accepts day-of-year from 0 to 365 */
+
+	token = args[7];
+	if (token == NULL) {
+		tm.tm_yday = 0;
+	}
+	else {
```
Comment (sunghan-chang) [inline apps/system/utils/kdbg_date.c]: Put this line at above (same line with closing bracket of if). This is a violation of coding rule.

File: apps/system/utils/kdbg_date.c [hunk 17]
```diff
@@ -205,6 +228,51 @@ static inline int date_settime(int argc, char **args)
...
+		tm.tm_yday = (int)result;
+	}
+
+	/* Set the daylight savings flag value */
+
+	token = args[8];
+	if (token == NULL) {
+		tm.tm_isdst = 0;
+	}
+	else {
```
Comment (sunghan-chang) [inline apps/system/utils/kdbg_date.c]: same as previous

File: apps/system/utils/kdbg_date.c [hunk 18]
```diff
@@ -235,9 +303,10 @@ int kdbg_date(int argc, char **args)
 
 	if (argc == 1) {
 		ret = date_showtime();
-	} else if (!strcmp(args[1], "-s") && argc == 6) {
+	} else if (!strcmp(args[1], "-s") && (argc >= 7 && argc <= 9)) {
```
Comment (sunghan-chang) [inline apps/system/utils/kdbg_date.c]: As I mentioned before, this is wrong when LIBC_LOCALTIME is disabled.
```
#if !(defined(CONFIG_LIBC_LOCALTIME) || defined(CONFIG_TIME_EXTENDED))
} else if (!strcmp(args[1], "-s") && (argc == 6)) {
#else
} else if (!strcmp(args[1], "-s") && (argc >= 7 && argc <= 9)) {
#endif
```

Comment (thapav) [inline apps/system/utils/kdbg_date.c]: oh okay, i will do that

**Comment by JunhwanPark** [comment]
Hi @junmin-kim , @thapav ,
Thanks for implementing it. I updated to the latest source.
I set it to `artik053/nettest` and enabled only `CONFIG_LIBC_LOCALTIME`. And I tested it like this:
```
TASH>> date -s Apr 21 10:35:22 1991 Sun 175 1
```
But I did not get the result I wanted. 
```
TASH>> date
Jan 01 00:00:00 1970 Thu
```
And the user must set all day of week, day of year, and daylight saving time? 
This is so confusing.
